File: Tuner/server/src/suggest.ts
// Turn the per-corner, per-phase diagnosis (diagnosis.ts) into signed setup-slider
// suggestions. Pure functions, no state.
//
// The honest model (see vault: BoxBox Tuner): direction is deterministic, magnitude
// is not. The SIGN of each change falls straight out of the standard tuning table;
// the number of clicks has no published value, so we seed it from hand-authored
// priors and let the online loop (2d-3) replace each prior with a measured gain.
// Every suggestion here is therefore "prior" confidence until that loop runs.
//
// Anchoring (the "hybrid" choice): the slip balance carries a baseline understeer
// bias (~+2 deg even on a neutral car), so a raw absolute reading would always say
// "add front wing". We handle it two ways:
//   - mid-corner (the setup axis): subtract a baseline-bias constant before judging
//     direction. That constant is the weakest assumption in the tool and is what
//     the loop refines first.
//   - traction (the exit axis) and entry: phase-RELATIVE, so they need no absolute
//     anchor. Traction is how far the rear gives up under power vs. a small target;
//     entry is how the braking phase compares to mid.
import type { CarSetupEntry } from "../../../shared/parser/index.ts";
import type { CornerDiagnosis } from "./diagnosis.ts";

// Levers we advise on: the few dominant ones the design says converge fast (wings,
// ARB, diff, brake bias). Fine params (camber, toe, ride height, pressures,
// suspension) need far more data than a handful of laps give, so they get no
// suggestion - their rows stay blank rather than show a guess dressed as advice.
export type SuggestKey =
  | "frontWing"
  | "rearWing"
  | "onThrottle"
  | "offThrottle"
  | "frontAntiRollBar"
  | "rearAntiRollBar"
  | "brakeBias";

export type Confidence = "prior" | "forming" | "measured";

export interface SetupSuggestion {
  key: SuggestKey;
  delta: number; // signed, in native step units (1 click = 1 here for every lever)
  confidence: Confidence;
  basis: string; // short reason, e.g. "mid understeer"
}

export interface SetupAdvice {
  headline: string;
  suggestions: SetupSuggestion[];
}

// --- Tunable priors (all guesses, refined by the loop) ------------------------
// Degrees are friendlier to reason about than radians; convert once.
const DEG = Math.PI / 180;

// The central assumption: the slip-balance reading of a neutral car. The ~+2.5-2.9
// deg means seen on real captures conflate the sensor's inherent front-slip bias
// with the car's actual understeer, so the true neutral sits below them. We set a
// modest 1.0 deg: low enough that a genuinely understeering baseline gets a mild
// dominant-lever nudge (so the driver has a change to apply and the loop has a
// gain to measure), high enough not to scream. 2d-3 measures the real per-car
// neutral and retires this constant - it is the weakest assumption in the tool.
const BASELINE_BIAS = 1.0 * DEG;
const MID_DEADBAND = 0.5 * DEG; // ignore effective understeer smaller than this

// Driver balance preference shifts the target the suggestions aim for, away from
// neutral. Two skilled players were opposite (one wants a touch of understeer for
// consistency, one wants it loose), so the target is personal, not universal (see
// vault: BoxBox Tuner Balance Preferences). The dial is normalized -1..+1: +1 =
// prefers understeer/stable, -1 = prefers oversteer/loose. At full deflection it
// moves the target by this much.
const PREF_RANGE = 2.0 * DEG;
const clampPref = (p: number): number => Math.max(-1, Math.min(1, p));
const EXIT_TARGET = 0.5 * DEG; // exit balance below this (on power) = rear at the limit
const ENTRY_DEADBAND = 1.0 * DEG; // entry is the noisiest phase; demand a clear delta

const POWER_THROTTLE = 0.5; // an exit aggregate above this mean throttle is "on power"
const MIN_SEEN = 2; // only corners confirmed on >= 2 laps feed car-level advice

// Per-lever response to each balance axis, as signed clicks per radian of that
// axis's excess. Sign encodes the tuning table; magnitude is the prior gain.
// A lever may answer to more than one axis (rear wing trades understeer against
// power-oversteer); contributions sum.
interface LeverGain {
  key: SuggestKey;
  axis: "mid" | "traction" | "entry";
  perRad: number;
}
const GAINS: LeverGain[] = [
  // Mid-corner understeer (signed: + understeer, - oversteer):
  { key: "frontWing", axis: "mid", perRad: 60 }, //   understeer -> +front wing
  { key: "rearWing", axis: "mid", perRad: -20 }, //   understeer -> -rear wing
  { key: "frontAntiRollBar", axis: "mid", perRad: -30 }, // understeer -> softer front ARB
  { key: "offThrottle", axis: "mid", perRad: -25 }, // understeer -> less off-throttle diff
  // Power-oversteer on exit (traction, >= 0):
  { key: "onThrottle", axis: "traction", perRad: 40 }, // -> more on-throttle diff lock
  { key: "rearAntiRollBar", axis: "traction", perRad: -30 }, // -> softer rear ARB
  { key: "rearWing", axis: "traction", perRad: 25 }, // -> +rear wing
  // Entry (signed: + understeer-on-entry vs mid, - looser-on-entry):
  { key: "brakeBias", axis: "entry", perRad: -60 }, // entry understeer -> bias rearward
];

// Slider bounds (mirror Tuner/web presentation/setup.ts) so a delta never pushes a
// value past its range. Step is 1 native unit per click for all of these.
const BOUNDS: Record<SuggestKey, { min: number; max: number }> = {
  frontWing: { min: 0, max: 50 },
  rearWing: { min: 0, max: 50 },
  onThrottle: { min: 10, max: 100 },
  offThrottle: { min: 10, max: 100 },
  frontAntiRollBar: { min: 1, max: 21 },
  rearAntiRollBar: { min: 1, max: 21 },
  brakeBias: { min: 50, max: 70 },
};
const CAP: Record<SuggestKey, number> = {
  frontWing: 3,
  rearWing: 3,
  onThrottle: 5,
  offThrottle: 5,
  frontAntiRollBar: 3,
  rearAntiRollBar: 3,
  brakeBias: 3,
};

export interface BalanceRollup {
  midBalance: number | null; // radians, sample-weighted across seen>=2 corners
  exitBalance: number | null; // on-power exits only
  entryBalance: number | null;
  midSamples: number;
  exitSamples: number;
  entrySamples: number;
}

// Sample-weighted car-level mean of one phase across confirmed corners. `gate`
// optionally excludes phase aggregates (e.g. off-power exits) from the mean.
function weightedPhase(
  diag: CornerDiagnosis[],
  phase: "entry" | "mid" | "exit",
  gate?: (throttle: number, brake: number) => boolean,
): { mean: number | null; samples: number } {
  let sum = 0;
  let n = 0;
  for (const d of diag) {
    if (d.seen < MIN_SEEN) continue;
    const p = d[phase];
    if (!p) continue;
    if (gate && !gate(p.throttle, p.brake)) continue;
    sum += p.slipBalance * p.samples;
    n += p.samples;
  }
  return { mean: n > 0 ? sum / n : null, samples: n };
}

export function rollupDiagnosis(diag: CornerDiagnosis[]): BalanceRollup {
  const mid = weightedPhase(diag, "mid");
  const exit = weightedPhase(diag, "exit", (th) => th > POWER_THROTTLE);
  const entry = weightedPhase(diag, "entry");
  return {
    midBalance: mid.mean,
    exitBalance: exit.mean,
    entryBalance: entry.mean,
    midSamples: mid.samples,
    exitSamples: exit.samples,
    entrySamples: entry.samples,
  };
}

function clampDelta(key: SuggestKey, raw: number, current: number): number {
  const snapped = Math.round(raw);
  const capped = Math.max(-CAP[key], Math.min(CAP[key], snapped));
  const b = BOUNDS[key];
  // Don't suggest moving past the slider's range from where it sits now.
  const lo = b.min - current;
  const hi = b.max - current;
  return Math.max(lo, Math.min(hi, capped));
}

/**
 * Build the setup advice from the diagnosis and the current setup. Returns null
 * until there is enough to say anything (no confirmed corners with samples).
 */
export function suggestSetup(
  diag: CornerDiagnosis[],
  setup: CarSetupEntry,
  preference = 0,
): SetupAdvice | null {
  const roll = rollupDiagnosis(diag);
  if (roll.midSamples === 0 && roll.exitSamples === 0 && roll.entrySamples === 0) return null;

  // Axis excesses (radians). mid is bias-adjusted absolute, shifted by the driver's
  // balance preference; traction and entry are phase-relative so they need no
  // anchor and are preference-independent.
  const target = BASELINE_BIAS + clampPref(preference) * PREF_RANGE;
  const midExcess =
    roll.midBalance === null ? 0 : signedDeadband(roll.midBalance - target, MID_DEADBAND);
  const tractionExcess =
    roll.exitBalance === null ? 0 : Math.max(0, EXIT_TARGET - roll.exitBalance);
  const entryExcess =
    roll.entryBalance === null || roll.midBalance === null
      ? 0
      : signedDeadband(roll.entryBalance - roll.midBalance, ENTRY_DEADBAND);
  const excess = { mid: midExcess, traction: tractionExcess, entry: entryExcess };

  const basis = (axis: LeverGain["axis"]): string =>
    axis === "mid"
      ? midExcess > 0
        ? "mid understeer"
        : "mid oversteer"
      : axis === "traction"
        ? "power oversteer on exit"
        : entryExcess > 0
          ? "entry understeer"
          : "entry instability";

  // Sum each lever's per-axis contributions; remember the axis whose single
  // contribution was largest, to label the suggestion's basis.
  const raw = new Map<SuggestKey, { delta: number; topAxis: LeverGain["axis"]; topMag: number }>();
  for (const g of GAINS) {
    const contrib = excess[g.axis] * g.perRad;
    if (contrib === 0) continue;
    const prev = raw.get(g.key);
    if (!prev) {
      raw.set(g.key, { delta: contrib, topAxis: g.axis, topMag: Math.abs(contrib) });
    } else {
      prev.delta += contrib;
      if (Math.abs(contrib) > prev.topMag) {
        prev.topMag = Math.abs(contrib);
        prev.topAxis = g.axis;
      }
    }
  }

  const suggestions: SetupSuggestion[] = [];
  for (const [key, { delta, topAxis }] of raw) {
    const d = clampDelta(key, delta, setup[key]);
    if (d === 0) continue;
    suggestions.push({ key, delta: d, confidence: "prior", basis: basis(topAxis) });
  }
  // Order by absolute magnitude so the dominant lever reads first.
  suggestions.sort((a, b) => Math.abs(b.delta) - Math.abs(a.delta));

  return { headline: headlineFor(roll, midExcess, tractionExcess), suggestions };
}

function signedDeadband(x: number, band: number): number {
  if (x > band) return x - band;
  if (x < -band) return x + band;
  return 0;
}

// Headline phrased relative to the driver's target, since the preference shifts it:
// "understeer" here means "more understeer than you want", not an absolute.
function headlineFor(roll: BalanceRollup, midExcess: number, tractionExcess: number): string {
  const parts: string[] = [];
  if (midExcess > 0.5 * DEG) parts.push("understeer vs your target");
  else if (midExcess < -0.5 * DEG) parts.push("looser than your target");
  else parts.push("on your target mid-corner");
  if (tractionExcess > 0.5 * DEG) parts.push("rear loose on power");
  return parts.join(", ");
}

File: app/src-tauri/src/tuner/suggest.rs
//! Turn the per-corner, per-phase diagnosis into signed setup-slider suggestions.
//! Pure functions, no state. Ported from `Tuner/server/src/suggest.ts`.
//!
//! The honest model: the SIGN of each change falls straight out of the standard
//! tuning table; the number of clicks is seeded from hand-authored priors and the
//! online loop replaces each prior with a measured gain. Mid-corner is
//! bias-adjusted absolute; traction (exit) and entry are phase-relative.

use std::collections::HashMap;

use serde::{Deserialize, Serialize};

use super::diagnosis::CornerDiagnosis;
use super::estimator::{lever_channel, Channel, LearnedGain};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SuggestKey {
    FrontWing,
    RearWing,
    OnThrottle,
    OffThrottle,
    FrontAntiRollBar,
    RearAntiRollBar,
    BrakeBias,
}

impl SuggestKey {
    /// The stable string key used in the persisted profile. Matches the camelCase
    /// serde name so existing on-disk profiles keep loading (P3.4).
    pub fn key(self) -> &'static str {
        match self {
            SuggestKey::FrontWing => "frontWing",
            SuggestKey::RearWing => "rearWing",
            SuggestKey::OnThrottle => "onThrottle",
            SuggestKey::OffThrottle => "offThrottle",
            SuggestKey::FrontAntiRollBar => "frontAntiRollBar",
            SuggestKey::RearAntiRollBar => "rearAntiRollBar",
            SuggestKey::BrakeBias => "brakeBias",
        }
    }

    /// Parse a persisted key, or None if this build doesn't know the lever.
    pub fn from_key(s: &str) -> Option<Self> {
        Some(match s {
            "frontWing" => SuggestKey::FrontWing,
            "rearWing" => SuggestKey::RearWing,
            "onThrottle" => SuggestKey::OnThrottle,
            "offThrottle" => SuggestKey::OffThrottle,
            "frontAntiRollBar" => SuggestKey::FrontAntiRollBar,
            "rearAntiRollBar" => SuggestKey::RearAntiRollBar,
            "brakeBias" => SuggestKey::BrakeBias,
            _ => return None,
        })
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum Confidence {
    Prior,
    Forming,
    Measured,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SetupSuggestion {
    pub key: SuggestKey,
    pub delta: i32, // signed, in native step units (1 click = 1)
    pub confidence: Confidence,
    pub basis: String,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct SetupAdvice {
    pub headline: String,
    pub suggestions: Vec<SetupSuggestion>,
}

const DEG: f64 = std::f64::consts::PI / 180.0;
// Neutral preference targets a truly neutral car; the preset shifts the target a
// modest ±1° (Loose → ~1° oversteer, Stable → ~1° understeer). Previously a 1°
// baseline + 2° range put "Stable" at a 3° understeer target — six times the
// per-corner US threshold — so a plowing car read as "dialed in".
const BASELINE_BIAS: f64 = 0.0;
const MID_DEADBAND: f64 = 0.5 * DEG;
const PREF_RANGE: f64 = 1.0 * DEG;
// Absolute understeer ceiling: no preference makes heavy understeer acceptable.
// Past this, advice to reduce understeer is always surfaced so a heavily
// understeering car is never reported as on-target, whatever the preset.
const MID_UNDERSTEER_CEILING: f64 = 2.0 * DEG;
const EXIT_TARGET: f64 = 0.5 * DEG;
const ENTRY_DEADBAND: f64 = 1.0 * DEG;
const POWER_THROTTLE: f64 = 0.5;
const MIN_SEEN: u32 = 2;

fn clamp_pref(p: f64) -> f64 {
    p.clamp(-1.0, 1.0)
}

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum Axis {
    Mid,
    Traction,
    Entry,
}

struct LeverGain {
    key: SuggestKey,
    axis: Axis,
    per_rad: f64,
}

// The suggestion axis a learned channel maps to (exit channel = traction axis).
fn channel_axis(channel: Channel) -> Axis {
    match channel {
        Channel::Mid => Axis::Mid,
        Channel::Exit => Axis::Traction,
        Channel::Entry => Axis::Entry,
    }
}

fn gains_table() -> [LeverGain; 8] {
    use Axis::*;
    use SuggestKey::*;
    [
        // Mid-corner understeer (signed: + understeer, - oversteer):
        LeverGain {
            key: FrontWing,
            axis: Mid,
            per_rad: 60.0,
        },
        LeverGain {
            key: RearWing,
            axis: Mid,
            per_rad: -20.0,
        },
        LeverGain {
            key: FrontAntiRollBar,
            axis: Mid,
            per_rad: -30.0,
        },
        LeverGain {
            key: OffThrottle,
            axis: Mid,
            per_rad: -25.0,
        },
        // Power-oversteer on exit (traction, >= 0):
        LeverGain {
            key: OnThrottle,
            axis: Traction,
            per_rad: 40.0,
        },
        LeverGain {
            key: RearAntiRollBar,
            axis: Traction,
            per_rad: -30.0,
        },
        LeverGain {
            key: RearWing,
            axis: Traction,
            per_rad: 25.0,
        },
        // Entry (signed: + understeer-on-entry vs mid, - looser-on-entry):
        LeverGain {
            key: BrakeBias,
            axis: Entry,
            per_rad: -60.0,
        },
    ]
}

// Slider bounds so a delta never pushes a value past its range.
fn bounds(key: SuggestKey) -> (f64, f64) {
    match key {
        SuggestKey::FrontWing => (0.0, 50.0),
        SuggestKey::RearWing => (0.0, 50.0),
        SuggestKey::OnThrottle => (10.0, 100.0),
        SuggestKey::OffThrottle => (10.0, 100.0),
        SuggestKey::FrontAntiRollBar => (1.0, 21.0),
        SuggestKey::RearAntiRollBar => (1.0, 21.0),
        SuggestKey::BrakeBias => (50.0, 70.0),
    }
}

fn cap(key: SuggestKey) -> i32 {
    match key {
        SuggestKey::FrontWing | SuggestKey::RearWing => 3,
        SuggestKey::OnThrottle | SuggestKey::OffThrottle => 5,
        SuggestKey::FrontAntiRollBar | SuggestKey::RearAntiRollBar => 3,
        SuggestKey::BrakeBias => 3,
    }
}

#[derive(Debug, Clone, Copy)]
pub struct BalanceRollup {
    pub mid_balance: Option<f64>,
    pub exit_balance: Option<f64>,
    pub entry_balance: Option<f64>,
    pub mid_samples: u32,
    pub exit_samples: u32,
    pub entry_samples: u32,
}

// Sample-weighted car-level mean of one phase across confirmed corners.
fn weighted_phase(
    diag: &[CornerDiagnosis],
    phase: Axis, // Mid/Entry pass-through; exit gated separately by caller
    select: impl Fn(&CornerDiagnosis) -> Option<(f64, u32, f64)>, // (slipBalance, samples, throttle)
    power_gate: bool,
) -> (Option<f64>, u32) {
    let _ = phase;
    let mut sum = 0.0;
    let mut n: u32 = 0;
    for d in diag {
        if d.seen < MIN_SEEN {
            continue;
        }
        let Some((slip_balance, samples, throttle)) = select(d) else {
            continue;
        };
        if power_gate && throttle <= POWER_THROTTLE {
            continue;
        }
        sum += slip_balance * samples as f64;
        n += samples;
    }
    (if n > 0 { Some(sum / n as f64) } else { None }, n)
}

pub fn rollup_diagnosis(diag: &[CornerDiagnosis]) -> BalanceRollup {
    let (mid_balance, mid_samples) = weighted_phase(
        diag,
        Axis::Mid,
        |d| d.mid.map(|p| (p.slip_balance, p.samples, p.throttle)),
        false,
    );
    let (exit_balance, exit_samples) = weighted_phase(
        diag,
        Axis::Traction,
        |d| d.exit.map(|p| (p.slip_balance, p.samples, p.throttle)),
        true,
    );
    let (entry_balance, entry_samples) = weighted_phase(
        diag,
        Axis::Entry,
        |d| d.entry.map(|p| (p.slip_balance, p.samples, p.throttle)),
        false,
    );
    BalanceRollup {
        mid_balance,
        exit_balance,
        entry_balance,
        mid_samples,
        exit_samples,
        entry_samples,
    }
}

fn signed_deadband(x: f64, band: f64) -> f64 {
    if x > band {
        x - band
    } else if x < -band {
        x + band
    } else {
        0.0
    }
}

fn clamp_delta(key: SuggestKey, raw: f64, current: f64) -> i32 {
    let snapped = raw.round() as i32;
    let c = cap(key);
    let capped = snapped.clamp(-c, c) as f64;
    let (min, max) = bounds(key);
    let lo = min - current;
    let hi = max - current;
    capped.clamp(lo, hi) as i32
}

/// Build the setup advice from the diagnosis and the current setup. Returns None
/// until there is enough to say anything.
pub fn suggest_setup(
    diag: &[CornerDiagnosis],
    setup_value: impl Fn(SuggestKey) -> f64,
    preference: f64,
    gains: &HashMap<SuggestKey, LearnedGain>,
) -> Option<SetupAdvice> {
    let roll = rollup_diagnosis(diag);
    if roll.mid_samples == 0 && roll.exit_samples == 0 && roll.entry_samples == 0 {
        return None;
    }

    let target = BASELINE_BIAS + clamp_pref(preference) * PREF_RANGE;
    let mid_excess = match roll.mid_balance {
        None => 0.0,
        Some(m) => {
            let pref_excess = signed_deadband(m - target, MID_DEADBAND);
            // Absolute guard: once past the ceiling, demand at least the amount
            // over it (understeer is positive). Only ever raises the understeer
            // signal — never cancels a "looser than target" one.
            let over_ceiling = (m - MID_UNDERSTEER_CEILING).max(0.0);
            if over_ceiling > 0.0 {
                pref_excess.max(over_ceiling)
            } else {
                pref_excess
            }
        }
    };
    let traction_excess = match roll.exit_balance {
        None => 0.0,
        Some(e) => (EXIT_TARGET - e).max(0.0),
    };
    let entry_excess = match (roll.entry_balance, roll.mid_balance) {
        (Some(en), Some(mid)) => signed_deadband(en - mid, ENTRY_DEADBAND),
        _ => 0.0,
    };

    let excess = |axis: Axis| match axis {
        Axis::Mid => mid_excess,
        Axis::Traction => traction_excess,
        Axis::Entry => entry_excess,
    };

    let basis = |axis: Axis| -> String {
        match axis {
            Axis::Mid => {
                if mid_excess > 0.0 {
                    "mid understeer".into()
                } else {
                    "mid oversteer".into()
                }
            }
            Axis::Traction => "power oversteer on exit".into(),
            Axis::Entry => {
                if entry_excess > 0.0 {
                    "entry understeer".into()
                } else {
                    "entry instability".into()
                }
            }
        }
    };

    // Accumulate per-lever contributions, preserving first-seen order (matches the
    // TS Map insertion order) so the later stable sort is deterministic.
    struct Accum {
        key: SuggestKey,
        delta: f64,
        top_axis: Axis,
        top_mag: f64,
    }
    let mut raw: Vec<Accum> = Vec::new();

    for g in gains_table() {
        let learned = gains.get(&g.key);
        let primary_axis = channel_axis(lever_channel(g.key).0);
        let per_rad = match learned {
            Some(l) if l.magnitude.is_some() && g.axis == primary_axis => {
                g.per_rad.signum() * l.magnitude.unwrap()
            }
            _ => g.per_rad,
        };
        let contrib = excess(g.axis) * per_rad;
        if contrib == 0.0 {
            continue;
        }
        match raw.iter_mut().find(|a| a.key == g.key) {
            None => raw.push(Accum {
                key: g.key,
                delta: contrib,
                top_axis: g.axis,
                top_mag: contrib.abs(),
            }),
            Some(prev) => {
                prev.delta += contrib;
                if contrib.abs() > prev.top_mag {
                    prev.top_mag = contrib.abs();
                    prev.top_axis = g.axis;
                }
            }
        }
    }

    let mut suggestions: Vec<SetupSuggestion> = Vec::new();
    for a in &raw {
        let d = clamp_delta(a.key, a.delta, setup_value(a.key));
        if d == 0 {
            continue;
        }
        let confidence = gains
            .get(&a.key)
            .map(|g| g.confidence)
            .unwrap_or(Confidence::Prior);
        suggestions.push(SetupSuggestion {
            key: a.key,
            delta: d,
            confidence,
            basis: basis(a.top_axis),
        });
    }
    // Order by absolute magnitude so the dominant lever reads first (stable).
    suggestions.sort_by_key(|s| std::cmp::Reverse(s.delta.abs()));

    Some(SetupAdvice {
        headline: headline_for(mid_excess, traction_excess),
        suggestions,
    })
}

fn headline_for(mid_excess: f64, traction_excess: f64) -> String {
    let mut parts: Vec<&str> = Vec::new();
    if mid_excess > 0.5 * DEG {
        parts.push("understeer vs your target");
    } else if mid_excess < -0.5 * DEG {
        parts.push("looser than your target");
    } else {
        parts.push("on your target mid-corner");
    }
    if traction_excess > 0.5 * DEG {
        parts.push("rear loose on power");
    }
    parts.join(", ")
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tuner::diagnosis::{PhaseDiagnosis, PhaseTone};

    const STABLE: f64 = 1.0;

    // A confirmed corner with only a mid-phase reading, to isolate mid balance.
    fn mid_corner(slip_balance_rad: f64) -> CornerDiagnosis {
        CornerDiagnosis {
            id: 1,
            index: 1,
            apex_dist: 0.0,
            min_speed: 100.0,
            seen: MIN_SEEN,
            entry: None,
            mid: Some(PhaseDiagnosis {
                samples: 10,
                slip_balance: slip_balance_rad,
                understeer_angle: 0.0,
                throttle: 0.0,
                brake: 0.0,
                tone: PhaseTone::Understeer,
            }),
            exit: None,
        }
    }

    #[test]
    fn heavy_understeer_is_advised_even_on_stable() {
        // ~3° understeer with the "Stable" preset must NOT read as dialed in:
        // that was the field bug (Stable's old 3° target swallowed the plow).
        let advice = suggest_setup(&[mid_corner(3.0 * DEG)], |_| 10.0, STABLE, &HashMap::new())
            .expect("advice once a corner is confirmed");
        assert!(
            advice.headline.contains("understeer vs your target"),
            "heavy understeer should flag understeer, got: {}",
            advice.headline
        );
        assert!(
            !advice.suggestions.is_empty(),
            "heavy understeer on Stable should still suggest a fix"
        );
    }

    #[test]
    fn mild_understeer_stays_on_target_on_stable() {
        // ~1° understeer == the Stable target: a planted car the driver asked for,
        // so no change is demanded (the preset is respected, just not to excess).
        let advice = suggest_setup(&[mid_corner(1.0 * DEG)], |_| 10.0, STABLE, &HashMap::new())
            .expect("advice");
        assert!(
            advice.headline.contains("on your target"),
            "mild understeer should read on-target at Stable, got: {}",
            advice.headline
        );
        assert!(
            advice.suggestions.is_empty(),
            "mild understeer at Stable should not demand a change"
        );
    }
}
